File: its-my-training-day/leetcode-top-75/questions/matrix/matrix-backtracking/word-search/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

bool wordSearch(std::vector<std::vector<char>>& board, const std::string& word) {
    int rows = static_cast<int>(board.size());
    int cols = static_cast<int>(board[0].size());
    std::function<bool(int, int, int)> dfs = [&](int r, int c, int idx) {
        if (idx == static_cast<int>(word.size())) return true;
        if (r < 0 || r >= rows || c < 0 || c >= cols) return false;
        if (board[r][c] != word[idx]) return false;
        char temp = board[r][c];
        board[r][c] = '#';
        bool found = dfs(r + 1, c, idx + 1) || dfs(r - 1, c, idx + 1) ||
                     dfs(r, c + 1, idx + 1) || dfs(r, c - 1, idx + 1);
        board[r][c] = temp;
        return found;
    };
    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            if (dfs(r, c, 0)) return true;
        }
    }
    return false;
}

File: its-my-training-day/leetcode-top-75/questions/matrix/matrix-backtracking/word-search/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def word_search(board: List[List[str]], word: str) -> bool:
    rows = len(board)
    cols = len(board[0])

    def dfs(r: int, c: int, idx: int) -> bool:
        if idx == len(word):
            return True
        if r < 0 or r >= rows or c < 0 or c >= cols:
            return False
        if board[r][c] != word[idx]:
            return False
        temp = board[r][c]
        board[r][c] = "#"
        found = (
            dfs(r + 1, c, idx + 1)
            or dfs(r - 1, c, idx + 1)
            or dfs(r, c + 1, idx + 1)
            or dfs(r, c - 1, idx + 1)
        )
        board[r][c] = temp
        return found

    r = 0
    while r < rows:
        c = 0
        while c < cols:
            if dfs(r, c, 0):
                return True
            c += 1
        r += 1
    return False
